reframe/listener.go

```diff
@@ -0,0 +1,507 @@
+}
+
+// NewReframeListenerWithNonceGen creates a reframe listener and initialises its state from the provided datastore.
+func New(ctx context.Context, engine provider.Interface, cidTtl time.Duration, chunkSize int, providerId string, addresses []string, ds datastore.Datastore, nonceGen func() []byte) (*ReframeListener, error) {
+	if nonceGen == nil {
+		nonceGen = defaultNonceGen
+	}
+
+	indexProvider := &ReframeListener{
+		engine:                 engine,
+		cidTtl:                 cidTtl,
+		chunkSize:              chunkSize,
+		currentChunk:           &cidsChunk{Cids: make(map[cid.Cid]struct{})},
+		listNodeByCid:          make(map[cid.Cid]*list.Element),
+		chunkByContextId:       make(map[string]*cidsChunk),
+		chunkByCid:             make(map[cid.Cid]*cidsChunk),
+		dsWrapper:              &dsWrapper{ds: ds},
+		nonceGen:               nonceGen,
+		lastSeenProviderInfo:   &peer.AddrInfo{},
+		configuredProviderInfo: nil,
+		nodesLl:                list.New(),
+		stats:                  stats{},
+	}
+
+	lister := &ReframeMultihashLister{
+		CidFetcher: func(contextID []byte) (map[cid.Cid]struct{}, error) {
+			ctxIdStr := contextIDToStr(contextID)
+			// that can be our current chunk that is not in the index yet
+			if ctxIdStr == contextIDToStr(indexProvider.currentChunk.ContextID) {
+				return indexProvider.currentChunk.Cids, nil
+			}
+			chunk := indexProvider.chunkByContextId[ctxIdStr]
+			if chunk == nil {
+				return nil, fmt.Errorf("multihasLister couldn't find a chunk for contextID %s", contextIDToStr(contextID))
+			}
+			return chunk.Cids, nil
+		},
+	}
+	engine.RegisterMultihashLister(lister.MultihashLister)
+
+	log.Info("Initialising from the datastore")
+	err := indexProvider.dsWrapper.initialiseFromTheDatastore(ctx, func(n *cidNode) {
+		listElem := indexProvider.nodesLl.PushFront(n)
+		indexProvider.listNodeByCid[n.c] = listElem
+	}, func(chunk *cidsChunk) {
+		indexProvider.chunkByContextId[contextIDToStr(chunk.ContextID)] = chunk
+		for k := range chunk.Cids {
+			indexProvider.chunkByCid[k] = chunk
+		}
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("Loaded up %d cids and %d chunks from the datastore.", len(indexProvider.listNodeByCid), len(indexProvider.chunkByContextId))
+
+	if providerId != "" {
+		p, err := peer.Decode(providerId)
+		if err != nil {
+			return nil, err
+		}
+
+		maddrs := make([]multiaddr.Multiaddr, len(addresses))
+		for i, s := range addresses {
+			a, err := multiaddr.NewMultiaddr(s)
+			if err != nil {
+				return nil, err
+			}
+			maddrs[i] = a
+		}
+
+		indexProvider.configuredProviderInfo = &peer.AddrInfo{
+			ID:    p,
+			Addrs: maddrs,
+		}
+	}
+
+	return indexProvider, nil
+}
+
+func (ReframeListener) GetIPNS(ctx context.Context, id []byte) (<-chan client.GetIPNSAsyncResult, error) {
+	log.Warn("Received unsupported getIPNS request")
+	ch := make(chan client.GetIPNSAsyncResult)
+	go func() {
+		// Not implemented
+		ch <- client.GetIPNSAsyncResult{Record: nil}
+		close(ch)
```

Comment:
Note that it is only necessary to close a channel if the receiver is looking for a close. Closing the channel is a control signal on the channel indicating that no more data follows. A channel being open does not constitute a reference to the channel object, and so does not keep it from being garbage collected. It is not like a file handle.